[Lim Zheng Ting] iP

AL-ZT opened the pull request:
# Duke v0.1
> “I don't know what to say so I'm linking my github here” – Zheng Ting ([source](https://github.com/AL-ZT))

Duke frees your mind of having to remember things you need to do. It's,

- text-based
- Not so easy to learn
- SLOW SUPER SLOW to use

All you need to do is,

1. download it from [here](https://github.com/AL-ZT/ip/releases/tag/v0.1).
2. download the jar file and run it with `java -jar <file-directory>/Duke-0.1.jar`.
3. add your tasks.
4. question your existence😉

And it is **FREE!**

Features:

- [x] Managing tasks, deadlines, events
- [x] Find a task
- [ ] GUI (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```Java
public class Main {
    public static void main(String[] args) {
        new Duke("./data/duke.txt").run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

AL-ZT committed:
Duke Level-0 + updated gitignore for Java class files

AL-ZT committed:
Duke Level-1 - Echo implementation

AL-ZT committed:
Duke Level-2: Add, List

AL-ZT committed:
Duke Level-3: Mark as Done

AL-ZT committed:
Duke Level-4: ToDo, Event, Deadline

AL-ZT committed:
Duke A-TextUiTesting

AL-ZT committed:
Duke Level-5: Minimal Error Handling

AL-ZT committed:
Duke Level-6: Delete Functionality

AL-ZT committed:
Minor customisations to bot

AL-ZT committed:
Duke Level-7: Save & Load Task data

AL-ZT committed:
Merge branch 'branch-Level-7'

AL-ZT committed:
Duke Level-8: Deadline dates with proper formatting

AL-ZT committed:
Merge branch 'branch-Level-8'

AL-ZT committed:
Almost done A-MoreOOP

AL-ZT committed:
Duke A-MoreOOP: Minimal OOP for Ui, Parser, Storage and TaskList classes

AL-ZT committed:
Merge: branch-A-MoreOOP

AL-ZT committed:
Merge branch 'branch-A-MoreOOP'

AL-ZT committed:
Divide classes into Packages

Classes can be categorised to avoid cluttering.

Packages and subpackages are introduced to easily allow developers to identify its controls to the application.

Let's further package the classes should the need arises.

AL-ZT committed:
Better exception handling

AL-ZT committed:
Merge branch 'branch-A-Packages'

AL-ZT committed:
Merge branch 'add-gradle-support'

AL-ZT committed:
JUnit test cases of Parser and TaskList classes

Unit testing is essential for maintaining accuracy of program throughout its versions.

Introducing JUnit test cases will ensure it aligns with user's expected results.

More test cases will be added in the future.

AL-ZT committed:
Merge branch 'branch-A-Junit'

AL-ZT committed:
Minor code changes

AL-ZT committed:
Minor changes to build name

AL-ZT committed:
Improved coding standards

AL-ZT committed:
Add JavaDocs for clarity

Provides documentation to programmers without the need to read the code itself.

JavaDocs provides documentation to increase readability of non-private constructors and functions.

JavaDocs will be continously added henceforth.

AL-ZT committed:
Duke Level-9: Find keyword

Give users a way to find a task by searching for a keyword.

AL-ZT committed:
Merge branch 'branch-A-JavaDoc'

AL-ZT committed:
Merge branch 'branch-A-CodingStandard'

AL-ZT committed:
Update TaskList::find with JavaDoc

AL-ZT committed:
Merge branch 'branch-Level-9'

AL-ZT committed:
Use checkStyle to detect coding style violations

Learn the different conventions of Java coding standards by enforcing it with checkStyle.

checkStyle offers automated style checking for every Java Source file, including Unit Tests.

AL-ZT committed:
Merge branch 'branch-A-CheckStyle'

AL-ZT committed:
Barely working GUI, needs more refactoring

AL-ZT committed:
Add GUI to Duke ChatBot

Command-Line Interface is not appropriate for an average user.

GUI allows cleaner interfaces to interact with and display relevant information.

Let's work on improving the User Interfaces.

AL-ZT committed:
Used Varargs

Pluiexo commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,55 @@
+    protected Command(String cmd, String[] args) {
+        this.command = cmd;
+        this.arguments = args;
+    }
```
Isn't it better to omit the this keyword because it is inherently understood where the command object belongs in this context?(This applies to other sections in the program as well)

AL-ZT replied:
Will do, thanks! It's a force of habit of mine.

Pluiexo commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,68 @@
+ * Bare task that has a task name and a boolean condition on whether the task is done.
+ */
+public class Task {
+    protected String taskName;
```
Having task in task name seems abit redundant, wouldn't it be better to shorten it to name?

Pluiexo commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,158 @@
+     * @param tasks <code>ArrayList</code> of tasks.
+     */
+    protected TaskList(ArrayList<Task> tasks) {
+        this.tasks = tasks;
```
This is a good example of when to keep the this keyword! From the previous comment message

Pluiexo reviewed (commented):
Overall, good formatting! Class names, variables names are up to standard and they convey a good message of what they represent. Although they may be some areas for improvement in some difficult areas of the formatting, its still good effort nonetheless!

Pluiexo commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,101 @@
+    protected void save(ArrayList<Task> tl) {
+        try {
+            this.data = new FileWriter(this.filePath);
+            tl.forEach(t -> {
+                try {
+                    if (t instanceof Todo) {
+                        this.data.write(String.format("T|%s|%s\n", t.getTaskName(), t.isDone()));
+                    } else if (t instanceof Deadline) {
+                        this.data.write(String.format("D|%s|%s|%s\n",
+                                t.getTaskName(), t.isDone(), (
+                                        (Deadline) t).getBy().format(DateTimeFormatter.ISO_LOCAL_DATE)
+                                )
+                        );
+                    } else if (t instanceof Event) {
+                        this.data.write(String.format("E|%s|%s|%s|%s\n",
+                                t.getTaskName(), t.isDone(), ((Event) t).getFrom(), ((Event) t).getTo()));
+                    }
+                } catch (IOException e) {
+                    System.out.println(e.getMessage());
+                }
+            });
```
Wow this is really difficult to read, the level of indentation seems to be off and the brackets are not leveled properly! May I suggest reconfiguring your wrapping settings in your IDE auto formatter if you are using one. Also nits tl could be better phrased into its full name for better readability in this case to differentiate between t of task

AL-ZT replied:
Thanks for your suggestion, I think I was merely following the checkstyle back then. If you don't mind me asking, where would the text wrap setting be in the IDE for this particular case? Much appreciated

Pluiexo replied:
If you are using IntelliJ -> settings -> code style -> java -> wrapping and braces. You can tweak the settings there to your preference!

AL-ZT replied:
Awesome, thanks a lot!

Pluiexo reviewed (commented):
Change selected lines

AL-ZT committed:
Merge branch 'branch-Level-10'

AL-ZT committed:
Use Assertions

Programmers' assumption with the code could cause some unexpected values to occur throughout the application.

Assertions are used to document important assumptions that should hold at various points in the code.

Let's try to use more assertions to explicitly lay out the constraints of variables if it isn't clear.

AL-ZT committed:
Improve Code Quality

Refactoring code improvews the readability of the code.

Extract long methods, avoid complicated expressions and magic numbers, and integrated SLAP into codebase.

Let's continue to further improve on existing and future lines of code.

AL-ZT committed:
Additional improvements to code readability

AL-ZT committed:
Merge pull request #1 from AL-ZT/branch-A-CodeQuality

Merge branch-A-CodeQuality

AL-ZT committed:
Merge pull request #2 from AL-ZT/branch-A-Assertions

Use Assertions

AL-ZT committed:
Trying Github Actions to set up CI

AL-ZT committed:
Fixed checkstyle error for CI

AL-ZT committed:
Add User Guide to Miku Chatbot

User guide allows quick and easy steps to get the application running on the other end.

This guide provides users with command details as well as its usage.

AL-ZT committed:
checkstyle error fix

AL-ZT committed:
Create _config.yml

AL-ZT committed:
Miku Loan Functionality Extension

Users may need to keep track of their loans, given or taken, to people in this general purpose app.

Provide user with a loan tracking functionality.

AL-ZT committed:
Modify README

AL-ZT committed:
Merge branch 'branch-BCD-Extension'

AL-ZT committed:
Update README.md
